File: scripts/redact.py
#!/usr/bin/env python3
"""Shared secret-redaction helpers for tool harness and evidence providers."""

from __future__ import annotations

import re


# ---------------------------------------------------------------------------
# Patterns that match common credential / token formats.
# ---------------------------------------------------------------------------

# GitHub personal access tokens (classic & fine-grained)
_RE_GHP = re.compile(r"ghp_[A-Za-z0-9]{30,}")
_RE_GITHUB_PAT = re.compile(r"github_pat_[A-Za-z0-9_]{20,}")

# Bearer / Basic auth headers and inline tokens
_RE_BEARER = re.compile(r"Bearer\s+[A-Za-z0-9._-]{20,}", re.IGNORECASE)
_RE_BASIC = re.compile(r"Basic\s+[A-Za-z0-9+/=]{20,}", re.IGNORECASE)

# Generic key=value / key: value patterns (common in logs, env dumps, configs)
_RE_KV_SECRET = re.compile(
    r"(?i)(api[_-]?key|token|password|secret|access[_-]?key|auth[_-]?token)"
    r"\s*[:=]\s*['\"]?[^\s'\"]{8,}"
)

# AWS-style access keys
_RE_AWS_KEY = re.compile(r"AKIA[0-9A-Z]{16}")

# Kubernetes / kubeconfig credentials. Only credential-bearing keys are
# matched: the old (server|username|...) form redacted every ordinary
# `server:` line in YAML evidence output, destroying legitimate review
# context for zero secrecy gain (server URLs and usernames are not secrets).
_RE_KUBE_CRED = re.compile(
    r"(?i)(password|client-certificate-data|client-key-data|"
    r"certificate-authority-data|bearer[_-]?token)\s*:\s*\S+"
)


def mask_secrets(text: str | None) -> str:
    """Return *text* with credential-like values replaced by ``[REDACTED]``.

    This is best-effort heuristic redaction; it will not catch every
    possible secret format and may occasionally false-positive on
    non-secret strings that happen to look similar.
    """
    if not text:
        return text or ""

    redacted = text

    for pattern in (
        _RE_GHP,
        _RE_GITHUB_PAT,
        _RE_BEARER,
        _RE_BASIC,
        _RE_KV_SECRET,
        _RE_AWS_KEY,
        _RE_KUBE_CRED,
    ):
        redacted = re.sub(pattern, "[REDACTED]", redacted)

    return redacted


def mask_and_truncate(text: str | None, max_bytes: int) -> tuple[str, bool]:
    """Redact secrets then truncate to *max_bytes*.

    Returns ``(masked_text, was_truncated)``.
    Truncation is performed **after** masking so the byte-length of the
    output reflects the redacted content (which may be shorter or longer
    than the original).
    """
    masked = mask_secrets(text)
    raw = masked.encode("utf-8", errors="replace")
    if len(raw) <= max_bytes:
        return masked, False

    clipped = raw[:max_bytes].decode("utf-8", errors="replace")
    return clipped + "\n[truncated]", True

File: src/context/redact.ts
/** Shared secret-redaction for v3 context producers (#675): a verbatim
 * TypeScript port of `scripts/redact.py`'s `redact_text`. The pr-thread and
 * related-code builders feed untrusted bodies and grep snippets through this
 * before anything reaches the corpus, exactly like their v2 counterparts.
 * The patterns, their application order, and the `[REDACTED]` marker are
 * contractual (v2/v3 parity compares the redacted bytes), so this module must
 * stay in lockstep with the Python original — it is a normalization seam, not
 * a security-policy one: no network, no process, no policy decisions. */

const REDACTED = "[REDACTED]";

// In application order (same as scripts/redact.py). Each pattern replaces
// every occurrence; later patterns see the already-redacted text.
const MASKERS: readonly RegExp[] = [
  // GitHub personal access tokens (classic & fine-grained)
  /ghp_[A-Za-z0-9]{30,}/g,
  /github_pat_[A-Za-z0-9_]{20,}/g,
  // Bearer / Basic auth headers and inline tokens
  /Bearer\s+[A-Za-z0-9._-]{20,}/gi,
  /Basic\s+[A-Za-z0-9+/=]{20,}/gi,
  // Generic key=value / key: value patterns (logs, env dumps, configs)
  /(api[_-]?key|token|password|secret|access[_-]?key|auth[_-]?token)\s*[:=]\s*['"]?[^\s'"]{8,}/gi,
  // AWS-style access keys
  /AKIA[0-9A-Z]{16}/g,
  // Kubernetes / kubeconfig credentials (credential-bearing keys only — the
  // old (server|username|...) form destroyed ordinary `server:` YAML context)
  /(password|client-certificate-data|client-key-data|certificate-authority-data|bearer[_-]?token)\s*:\s*\S+/gi,
];

/** Return *text* with credential-like values replaced by `[REDACTED]`.
 * Best-effort heuristic redaction, exactly like the Python original. */
export function redactText(text: string | null | undefined): string {
  if (!text) return "";
  let redacted = text;
  for (const pattern of MASKERS) {
    redacted = redacted.replace(pattern, REDACTED);
  }
  return redacted;
}
